### PR TITLE
CP-23007: Implement CD change using fd over QMP

### PR DESCRIPTION
**Pull request by fillzero**
Signed-off-by: Liang Dai <liang.dai1@citrix.com>

### CODE REVIEW

**Comment by fillzero** [comment]
@mseri - Would you please review this PR? Thanks.

**Comment by mseri** [comment]
@fillzero I am a bit packed at the moment, I'd prefer if somebody with more knowledge of the context is reviewing it

**Comment by fillzero** [comment]
@mg12 Would you please have a look? Thanks.

**Comment by lindig** [comment]
I looked at this again and I believe the JSON decoding can be massively simplified by taking advantage of the combinators that `Yojson` provides. Below is just a sketch for the central `message_of_string` module:

```
let message_of str =
  let module Y = Yojson.Basic in
  let module U = Yojson.Basic.Util in
  let greeting json =
    let version x = json |> U.member x |> U.to_int in
    { minor   = version "minor"
    ; major   = version "major"
    ; micro   = version "micro"
    ; package = json |> U.member "package" |> U.to_string
    } in
  let event json =
    let ts x = json |> U.member "timestamp" |> U.member x |> U.to_int in
    { timestamp = (ts "seconds", ts "microseconds")
    ; event     = json |> U.member "event" |> U.to_string
    }
  in
    let json = Y.from_string str in
    match json |> U.keys with
    | ["QMP"]   -> json |> U.member "QMP"   |> greeting |> fun x -> Greeting x
    | ["event"] -> json |> U.member "event" |> event |> fun x -> Event x
    | _         -> Printf.sprintf "can't decode %s" str |> failwith
```

**Comment by fillzero** [comment]
@lindig Are you happy to have a look?
PR https://github.com/xapi-project/xenopsd/pull/362 depends on this.

**Comment by lindig** [comment]
There are changes coming to this code by @mg12 who is re-writing the JSON reading.

File: lib/qmp.ml [hunk 1]
```diff
@@ -188,6 +200,15 @@ let message_of_string x =
...
                              | `Assoc [ "name", `String x ] -> x
                              | _ -> failwith "assoc") list))
+      | `Assoc list when List.mem_assoc "fdset-id" list -> (
+        try
+          Success (id, Fd_info (
+            let fd = int (List.assoc "fd" list) in
+            let fdset_id = int (List.assoc "fdset-id" list) in
+            {fd; fdset_id}))
+        with e ->
+          Error(None, { cls = "JSONParsing"; descr = (Printf.sprintf "%s:%s" (Printexc.to_string e) x) })
```
Comment (mg12) [inline lib/qmp.ml]: I think we need a test for this failure in messages.ml

Comment (fillzero) [inline lib/qmp.ml]: Thanks, updated.

**Pull request merged**